xcm/xcm-emulator/example/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+			Weight::from_parts(110000010, 10000010)
+		);
+		ParachainA::execute_with(|| {
+			assert_ok!(PolkadotXcm::send_xcm(
+				Here,
+				MultiLocation::new(1, X1(Parachain(2))),
+				Xcm(vec![Transact {
+					origin_kind: OriginKind::SovereignAccount,
+					require_weight_at_most: 110_000_010.into(),
+					call: send_xcm_to_octopus.encode().into(),
+				}]),
+			));
+		});
+
+		ParachainB::execute_with(|| {
+			use test_runtime::{RuntimeEvent, System};
+			System::events().iter().for_each(|r| println!(">>> {:?}", r.event));
+
+			assert!(System::events().iter().any(|r| matches!(
+				r.event,
+				RuntimeEvent::PolkadotXcm(pallet_xcm::Event::Sent(_, _, _))
+			)));
+		});
+
+		ParachainC::execute_with(|| {
+			use test_runtime::{RuntimeEvent, System};
+			// execution would fail, but good enough to check if the message is received
+			System::events().iter().for_each(|r| println!(">>> {:?}", r.event));
+
+			assert!(System::events().iter().any(|r| matches!(
+				r.event,
+				RuntimeEvent::XcmpQueue(cumulus_pallet_xcmp_queue::Event::Fail { .. })
+			)));
+		});
+	}
+
+	#[test]
+	fn deduplicate_dmp() {
+		Network::reset();
+		KusamaNet::execute_with(|| {
+			assert_ok!(kusama_runtime::XcmPallet::force_default_xcm_version(
+				kusama_runtime::RuntimeOrigin::root(),
+				Some(3)
+			));
+		});
+
+		kusama_send_rmrk("Kusama", 2);
+		parachain_receive_and_reset_events(true);
+
+		// a different dmp message in same relay-parent-block allow execution.
+		kusama_send_rmrk("Polkadot", 1);
+		parachain_receive_and_reset_events(true);
+
+		// same dmp message with same relay-parent-block wouldn't execution
+		kusama_send_rmrk("Kusama", 1);
+		parachain_receive_and_reset_events(false);
+
+		// different relay-parent-block allow dmp message execution
+		KusamaNet::execute_with(|| kusama_runtime::System::set_block_number(2));
+
+		kusama_send_rmrk("Kusama", 1);
+		parachain_receive_and_reset_events(true);
+
+		// reset can send same dmp message again
+		Network::reset();
+		KusamaNet::execute_with(|| {
+			assert_ok!(kusama_runtime::XcmPallet::force_default_xcm_version(
+				kusama_runtime::RuntimeOrigin::root(),
+				Some(3)
+			));
+		});
+
+		kusama_send_rmrk("Kusama", 1);
+		parachain_receive_and_reset_events(true);
+	}
+
+	fn kusama_send_rmrk(msg: &str, count: u32) {
+		let remark = test_runtime::RuntimeCall::System(
+			frame_system::Call::<test_runtime::Runtime>::remark_with_event {
+				remark: msg.as_bytes().to_vec(),
+			},
+		);
+		KusamaNet::execute_with(|| {
+			for _ in 0..count {
+				assert_ok!(kusama_runtime::XcmPallet::send_xcm(
+					Here,
+					Parachain(1),
+					Xcm(vec![Transact {
+						origin_kind: OriginKind::SovereignAccount,
+						require_weight_at_most: Weight::from_parts(
+							INITIAL_BALANCE as u64,
+							1024 * 1024
+						),
+						call: remark.encode().into(),
+					}]),
+				));
+			}
+		});
+	}
+
+	fn parachain_receive_and_reset_events(received: bool) {
+		ParachainA::execute_with(|| {
```

Comment:
Why is this function specific to `ParachainA`?
```suggestion
	fn parachain_a_verify_remark_event_and_reset_all_events(received: bool) {
		ParachainA::execute_with(|| {
```

Reply:
Only because `kusama_send_rmrk` is currently sending only to `parachain_a`.